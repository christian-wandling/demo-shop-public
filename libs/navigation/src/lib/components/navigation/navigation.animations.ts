import { animate, style, transition, trigger } from '@angular/animations';

export const animateBackdrop = trigger('animateBackdrop', [
  transition(':enter', [style({ opacity: 0 }), animate('300ms linear', style({ opacity: 1 }))]),
  transition(':leave', [style({ opacity: 1 }), animate('300ms linear', style({ opacity: 0 }))]),
]);

export const animateSlideOver = trigger('animateSlideOver', [
  transition(':enter', [
    style({ transform: 'translateX(-100%)' }),
    animate('300ms ease-in-out', style({ transform: 'translateX(0)' })),
  ]),
  transition(':leave', [
    style({ transform: 'translateX(0)' }),
    animate('300ms ease-in-out', style({ transform: 'translateX(-100%)' })),
  ]),
]);
