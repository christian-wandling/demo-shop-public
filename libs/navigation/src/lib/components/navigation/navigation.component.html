@if (mobileMenuOpen()) {
<div
  [@animateBackdrop]
  (click)="mobileMenuOpen.set(false)"
  class="relative z-40 lg:hidden"
  role="dialog"
  aria-modal="true">
  <div class="fixed inset-0 bg-black bg-opacity-25" aria-hidden="true"></div>
</div>
<div
  [@animateSlideOver]
  class="absolute h-screen z-40 flex w-full max-w-xs flex-col overflow-y-auto bg-white pb-12 shadow-xl">
  <div class="sticky z-20 top-0 flex px-4 pb-2 pt-5">
    <button
      (click)="mobileMenuOpen.set(false)"
      type="button"
      class="relative opacity-90 -m-2 inline-flex items-center justify-center rounded-md p-2 text-gray-400 bg-white">
      <span class="absolute -inset-0.5"></span>
      <span class="sr-only">Close menu</span>
      <svg class="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" aria-hidden="true">
        <path stroke-linecap="round" stroke-linejoin="round" d="M6 18L18 6M6 6l12 12" />
      </svg>
    </button>
  </div>

  <!--       Links-->
  <div class="mt-2">
    <div class="space-y-10 px-4 pb-8">
      <div class="grid grid-cols-2 gap-x-4">
        <div class="group relative text-sm">
          <div class="aspect-h-1 aspect-w-1 overflow-hidden rounded-lg bg-gray-100 group-hover:opacity-75">
            <lib-image-placeholder text="N" />
          </div>
          <a class="mt-6 block font-medium text-gray-900">
            <span class="absolute inset-0 z-10" aria-hidden="true"></span>
            New Items
          </a>
          <p aria-hidden="true" class="mt-1">Shop now</p>
        </div>
        <div class="group relative text-sm">
          <div class="aspect-h-1 aspect-w-1 overflow-hidden rounded-lg bg-gray-100 group-hover:opacity-75">
            <lib-image-placeholder text="D" />
          </div>
          <a class="mt-6 block font-medium text-gray-900">
            <span class="absolute inset-0 z-10" aria-hidden="true"></span>
            Deals
          </a>
          <p aria-hidden="true" class="mt-1">Shop now</p>
        </div>
      </div>
      <div>
        @for (item of flyoutMenuItems; track $index) {
        <p id="{{ item.label }}-heading-mobile" class="font-medium text-gray-900">{{ item.label | titlecase }}</p>
        <ul
          role="list"
          [attr.aria-labelledby]="item.label + '-heading-mobile'"
          class="mt-6 mb-10 last:mb-0 flex flex-col space-y-6">
          @for (subItem of item.subItems; track $index) {
          <li class="flow-root">
            <a (click)="selectedMenuItem.set(item.label)" class="-m-2 block p-2 text-gray-500">{{
              subItem.label | titlecase
            }}</a>
          </li>
          }
          <li class="flow-root">
            <a (click)="selectedMenuItem.set('')" class="-m-2 block p-2 text-gray-500">Browse All</a>
          </li>
        </ul>
        }
      </div>
    </div>
  </div>

  <div class="space-y-6 border-t border-gray-200 px-4 py-6">
    @for (item of menuItems; track $index) {
    <div class="flow-root">
      <a [routerLink]="item.options.route" class="-m-2 block p-2 font-medium text-gray-900">
        {{ item.label | titlecase }}
      </a>
    </div>
    }
  </div>

  <lib-user-navigation isSmallScreen class="lg:hidden space-y-6 border-t border-gray-200 px-4 py-6">
  </lib-user-navigation>
</div>
}

<header class="relative bg-white">
  <nav aria-label="Top" class="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8">
    <div class="border-b border-gray-200">
      <div class="flex h-16 items-center">
        <button
          (click)="mobileMenuOpen.set(true)"
          type="button"
          class="relative rounded-md bg-white p-2 text-gray-400 lg:hidden">
          <span class="absolute -inset-0.5"></span>
          <span class="sr-only">Open menu</span>
          <svg
            class="h-6 w-6"
            fill="none"
            viewBox="0 0 24 24"
            stroke-width="1.5"
            stroke="currentColor"
            aria-hidden="true">
            <path stroke-linecap="round" stroke-linejoin="round" d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5" />
          </svg>
        </button>

        <!-- Logo -->
        <div class="ml-4 flex lg:ml-0 h-8 w-8 relative">
          <a routerLink="/">
            <span class="sr-only">Demo Shop</span>
            <img class="object-contain" ngSrc="icons/demo-shop.png" alt="Demo shop logo" fill="" />
          </a>
        </div>

        <!-- Flyout menus -->
        <div class="hidden lg:ml-8 lg:block lg:self-stretch">
          <div class="flex h-full space-x-8">
            <div (mouseleave)="flyoutMenuOpen.set(false)" class="flex space-x-8">
              @for (item of flyoutMenuItems; track $index) {
              <div class="menu-item flex h-full">
                <div class="relative flex">
                  <button
                    routerLink="products"
                    (mouseenter)="flyoutMenuOpen.set(true)"
                    type="button"
                    class="menu-item-button relative z-10 -mb-px flex items-center border-b-2 border-transparent pt-px text-sm font-medium text-gray-700 transition-colors duration-200 ease-out hover:text-gray-800"
                    aria-expanded="false">
                    {{ item.label | titlecase }}
                  </button>
                </div>

                @if (flyoutMenuOpen()) {
                <div [@animateFlyoutMenu] class="flyout-menu absolute inset-x-0 top-full text-sm text-gray-500">
                  <!-- Presentational element used to render the bottom shadow, if we put the shadow on the actual panel it pokes out the top, so we use this shorter element to hide the top of the shadow -->
                  <div class="absolute inset-0 top-1/2 bg-white shadow" aria-hidden="true"></div>

                  <div class="relative bg-white">
                    <div class="mx-auto max-w-7xl px-8">
                      <div class="grid grid-cols-2 gap-x-8 gap-y-10 py-16">
                        <div class="col-start-2 grid grid-cols-2 gap-x-8">
                          <div class="group relative text-base sm:text-sm">
                            <div
                              class="aspect-h-1 aspect-w-1 overflow-hidden rounded-lg bg-gray-100 group-hover:opacity-75">
                              <lib-image-placeholder text="N" />
                            </div>
                            <a class="mt-6 block font-medium text-gray-900">
                              <span class="absolute inset-0 z-10" aria-hidden="true"></span>
                              New Items
                            </a>
                            <p aria-hidden="true" class="mt-1">Shop now</p>
                          </div>
                          <div class="group relative text-base sm:text-sm">
                            <div
                              class="aspect-h-1 aspect-w-1 overflow-hidden rounded-lg bg-gray-100 group-hover:opacity-75">
                              <lib-image-placeholder text="D" />
                            </div>
                            <a class="mt-6 block font-medium text-gray-900">
                              <span class="absolute inset-0 z-10" aria-hidden="true"></span>
                              Deals
                            </a>
                            <p aria-hidden="true" class="mt-1">Shop now</p>
                          </div>
                        </div>
                        <div class="row-start-1 grid grid-cols-3 gap-x-8 gap-y-10 text-sm">
                          <div>
                            <p id="{{ item.label }}-heading" class="font-medium text-gray-900">
                              {{ item.label | titlecase }}
                            </p>
                            <ul
                              role="list"
                              [attr.aria-labelledby]="item.label + '-heading'"
                              class="mt-6 space-y-6 sm:mt-4 sm:space-y-4">
                              @for (subItem of item.subItems; track $index) {
                              <li class="flex">
                                <a
                                  (click)="setProductCategory(subItem.label)"
                                  class="hover:text-gray-800 cursor-pointer"
                                  >{{ subItem.label | titlecase }}</a
                                >
                              </li>
                              }
                            </ul>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                }
              </div>
              }
            </div>
            @for (item of menuItems; track $index) {
            <div class="menu-item flex h-full space-x-8">
              <a
                [routerLink]="item.options.route"
                class="menu-item-button flex items-center text-sm font-medium text-gray-700 hover:text-gray-800">
                {{ item.label | titlecase }}
              </a>
            </div>
            }
          </div>
        </div>

        <div class="ml-auto flex items-center">
          <lib-user-navigation class="hidden lg:flex lg:flex-1 lg:items-center lg:justify-end lg:space-x-6">
          </lib-user-navigation>
          <lib-product-search></lib-product-search>
          <lib-cart-icon></lib-cart-icon>
        </div>
      </div>
    </div>
  </nav>
</header>
