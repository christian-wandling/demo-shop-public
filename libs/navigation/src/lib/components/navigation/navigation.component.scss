:host {
  top: 0;
  position: sticky;
  z-index: 40;
}

.menu-item {
  &:hover {
    .menu-item-button {
      color: #4f46e5;
      border-color: #4f46e5;
    }
  }
}
