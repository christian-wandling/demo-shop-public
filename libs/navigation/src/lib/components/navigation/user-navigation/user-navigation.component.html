<ng-container *ngIf="isSmallScreen(); else largeScreen">
  @if (currentUser()) {
  <div class="flow-root">
    <a class="-m-2 block p-2 font-medium text-gray-900 cursor-pointer">
      {{ currentUser()?.firstname }}&nbsp;{{ currentUser()?.lastname }}
    </a>
  </div>
  <div class="flow-root">
    <a (click)="logout()" class="-m-2 block p-2 font-medium text-gray-900 cursor-pointer"> Sign out </a>
  </div>
  } @else {
  <div class="flow-root">
    <a (click)="login()" class="-m-2 block p-2 font-medium text-gray-900 cursor-pointer"> Sign in </a>
  </div>
  <div class="flow-root">
    <a (click)="register()" class="-m-2 block p-2 font-medium text-gray-900 cursor-pointer"> Register </a>
  </div>
  }
</ng-container>

<ng-template #largeScreen>
  @if (currentUser()) {
  <a class="text-sm font-medium text-gray-700 hover:text-gray-800 cursor-pointer">
    {{ currentUser()?.firstname }}&nbsp;{{ currentUser()?.lastname }}
  </a>
  <span class="h-6 w-px bg-gray-200" aria-hidden="true"></span>
  <a (click)="logout()" class="text-sm font-medium text-gray-700 hover:text-gray-800 cursor-pointer"> Sign out </a>
  } @else {
  <a (click)="login()" class="text-sm font-medium text-gray-700 hover:text-gray-800 cursor-pointer"> Sign in </a>
  <span class="h-6 w-px bg-gray-200" aria-hidden="true"></span>
  <a (click)="register()" class="text-sm font-medium text-gray-700 hover:text-gray-800 cursor-pointer"> Register </a>
  }
</ng-template>
