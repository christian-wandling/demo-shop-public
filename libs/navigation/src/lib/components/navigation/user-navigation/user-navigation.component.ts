import { booleanAttribute, ChangeDetectionStrategy, Component, inject, input } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AuthFacade } from '@demo-shop/auth';
import { UserFacade } from '@demo-shop/user';
import { CartFacade } from '@demo-shop/cart';

/**
 * Component that handles user navigation functionality including login, registration, and logout
 *
 * @example
 * <lib-user-navigation isSmallScreen>
 */
@Component({
  selector: 'lib-user-navigation',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './user-navigation.component.html',
  styleUrl: './user-navigation.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class UserNavigationComponent {
  /** Flag indicating if the current view is on a small screen */
  readonly isSmallScreen = input(false, { transform: booleanAttribute });

  readonly #authFacade = inject(AuthFacade);
  readonly #userFacade = inject(UserFacade);
  readonly currentUser = this.#userFacade.getCurrentUser();
  readonly #cartFacade = inject(CartFacade);

  /** Handles user login and loads shopping session */
  async login(): Promise<void> {
    await this.#authFacade.login();
    await this.#cartFacade.loadShoppingSession();
  }

  /** Handles user registration and loads shopping session */
  async register(): Promise<void> {
    await this.#authFacade.register();
    await this.#cartFacade.loadShoppingSession();
  }

  /** Handles user logout */
  async logout(): Promise<void> {
    await this.#authFacade.logout();
  }
}
