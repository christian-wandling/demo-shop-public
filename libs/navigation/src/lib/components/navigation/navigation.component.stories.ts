import { applicationConfig, Meta, moduleMetadata, StoryObj } from '@storybook/angular';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { provideRouter } from '@angular/router';
import { CartFacade, CartIconComponent } from '@demo-shop/cart';
import { NavigationComponent } from './navigation.component';
import { ProductFacade, ProductSearchComponent } from '@demo-shop/product';
import { UserNavigationComponent } from './user-navigation/user-navigation.component';
import { provideAnimations } from '@angular/platform-browser/animations';
import { RouteItem } from '../../models/route-item';
import { NavigationService } from '../../services/navigation.service';
import { AuthFacade, PermissionStrategy } from '@demo-shop/auth';
import { UserResponse } from '@demo-shop/api';
import { signal } from '@angular/core';
import { UserFacade } from '@demo-shop/user';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { faker } from '@faker-js/faker';
import { FormsModule } from '@angular/forms';

const user: UserResponse = {
  id: 1,
  email: faker.internet.email(),
  firstname: faker.person.firstName(),
  lastname: faker.person.lastName(),
  phone: faker.helpers.fromRegExp('+[0-9]{9}'),
};

const navItems = [
  new RouteItem('products', 101, {
    route: 'products',
  }),
];

const authItems = [
  new RouteItem('orders', 102, {
    route: 'orders',
    permissionStrategy: PermissionStrategy.AUTHENTICATED,
  }),
];

const navigationService = (items: RouteItem[]) => ({
  getFilteredItems: () => items,
});

const productFilter = signal({ name: '' });

const mockProductFacade = {
  getFilter: () => productFilter,
  setFilter: (filter: { name: string }) => productFilter.set(filter),
};

const mockUserFacade = (user?: UserResponse) => ({
  getCurrentUser: () => signal(user),
});

const mockAuthFacade = {
  login: () => {
    return;
  },
  logout: () => {
    return;
  },
  register: () => {
    return;
  },
};

const mockCartFacade = {
  loadShoppingSession: () => {
    return;
  },
  getItemCount: () => signal(5),
};

const meta: Meta<NavigationComponent> = {
  component: NavigationComponent,
  title: 'Navigation/NavigationComponent',
  decorators: [
    applicationConfig({
      providers: [
        provideAnimations(),
        provideRouter([{ path: '**', redirectTo: '' }]),
        { provide: AuthFacade, useValue: mockAuthFacade },
        { provide: CartFacade, useValue: mockCartFacade },
        { provide: ProductFacade, useValue: mockProductFacade },
      ],
    }),
    moduleMetadata({
      imports: [
        CommonModule,
        FormsModule,
        NgOptimizedImage,
        ProductSearchComponent,
        UserNavigationComponent,
        CartIconComponent,
      ],
    }),
  ],
  parameters: {
    controls: {
      exclude: ['#navigationService', 'menuItems', 'mobileMenuOpen', 'selectedMenuItem'],
    },
  },
};

export default meta;
type Story = StoryObj<NavigationComponent>;

export const Unauthenticated: Story = {
  decorators: [
    applicationConfig({
      providers: [
        { provide: UserFacade, useValue: mockUserFacade() },
        { provide: NavigationService, useValue: navigationService(navItems) },
      ],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const product = within(canvasElement).getByText('Products');
    const signIn = within(canvasElement).getByText('Sign in');
    const register = within(canvasElement).getByText('Register');
    const productSearch = within(canvasElement).getByAltText('search');
    expect(product).toBeInTheDocument();
    expect(signIn).toBeInTheDocument();
    expect(register).toBeInTheDocument();
    expect(productSearch).toBeInTheDocument();
  },
};

export const Authenticated: Story = {
  decorators: [
    applicationConfig({
      providers: [
        { provide: UserFacade, useValue: mockUserFacade(user) },
        { provide: NavigationService, useValue: navigationService([...navItems, ...authItems]) },
      ],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const product = within(canvasElement).getByText('Products');
    const order = within(canvasElement).getByText('Orders');
    const signOut = within(canvasElement).getByText('Sign out');
    const userName = within(canvasElement).getByText(`${user.firstname} ${user.lastname}`);
    const productSearch = within(canvasElement).getByAltText('search');
    expect(product).toBeInTheDocument();
    expect(order).toBeInTheDocument();
    expect(signOut).toBeInTheDocument();
    expect(userName).toBeInTheDocument();
    expect(productSearch).toBeInTheDocument();
  },
};
