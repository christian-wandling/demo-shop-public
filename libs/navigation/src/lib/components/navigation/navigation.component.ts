import { ChangeDetectionStrategy, Component, inject, signal } from '@angular/core';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { Router, RouterLink } from '@angular/router';
import { ImagePlaceholderComponent } from '@demo-shop/shared';
import { NavigationService } from '../../services/navigation.service';
import { NavigationType } from '../../enums/navigation-type';
import { RouteItem } from '../../models/navigation-item';
import { ProductFacade, ProductSearchComponent } from '@demo-shop/product';
import { animateBackdrop, animateFlyoutMenu, animateSlideOver } from './navigation.animations';
import { UserNavigationComponent } from './user-navigation/user-navigation.component';
import { CartIconComponent } from '@demo-shop/cart';

@Component({
  selector: 'lib-navigation',
  standalone: true,
  imports: [
    CommonModule,
    RouterLink,
    NgOptimizedImage,
    ImagePlaceholderComponent,
    ProductSearchComponent,
    UserNavigationComponent,
    CartIconComponent,
  ],
  templateUrl: './navigation.component.html',
  styleUrl: './navigation.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
  animations: [animateBackdrop, animateSlideOver, animateFlyoutMenu],
  host: {
    class: 'bg-white',
    '(window:resize)': 'this.mobileMenuOpen.set(false);',
  },
})
export class NavigationComponent {
  readonly mobileMenuOpen = signal(false);
  readonly selectedMenuItem = signal('products');
  readonly flyoutMenuOpen = signal(false);

  readonly #productFacade = inject(ProductFacade);
  readonly #router = inject(Router);
  readonly #navigationService = inject(NavigationService);

  readonly flyoutMenuItems = this.#navigationService.getFilteredItems(NavigationType.FLYOUT);
  readonly menuItems = this.#navigationService.getFilteredItems(NavigationType.ROUTE) as RouteItem[];

  setProductCategory(category: string): void {
    this.#productFacade.setFilter({
      categories: category,
    });

    this.#router.navigateByUrl('/products');
  }
}
