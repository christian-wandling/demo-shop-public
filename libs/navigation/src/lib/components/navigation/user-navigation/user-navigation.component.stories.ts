import { applicationConfig, Meta, StoryObj } from '@storybook/angular';
import { CommonModule } from '@angular/common';
import { importProvidersFrom, signal } from '@angular/core';
import { UserResponse } from '@demo-shop/api';
import { CartFacade } from '@demo-shop/cart';
import { UserNavigationComponent } from './user-navigation.component';
import { mockUser, UserFacade } from '@demo-shop/user';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { AuthFacade } from '@demo-shop/auth';

const mockUserFacade = (user?: UserResponse) => ({
  getCurrentUser: () => signal(user),
});

const mockAuthFacade = {
  login: () => {
    return;
  },
  logout: () => {
    return;
  },
  register: () => {
    return;
  },
};

const mockCartFacade = {
  loadShoppingSession: () => {
    return;
  },
};

const meta: Meta<UserNavigationComponent> = {
  component: UserNavigationComponent,
  title: 'Navigation/UserNavigationComponent',
  decorators: [
    applicationConfig({
      providers: [
        importProvidersFrom(CommonModule),
        { provide: AuthFacade, useValue: mockAuthFacade },
        { provide: CartFacade, useValue: mockCartFacade },
      ],
    }),
  ],
  argTypes: {
    isSmallScreen: {
      control: {
        type: 'boolean',
      },
    },
    login: { action: 'login', control: false },
    register: { action: 'register', control: false },
    logout: { action: 'logout', control: false },
  },
  parameters: {
    controls: {
      exclude: ['#authFacade', '#userFacade', 'currentUser', '#cartFacade'],
    },
  },
};
export default meta;
type Story = StoryObj<UserNavigationComponent>;

export const Unauthenticated: Story = {
  args: {
    isSmallScreen: false,
  },
  decorators: [
    applicationConfig({
      providers: [{ provide: UserFacade, useValue: mockUserFacade() }],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const signIn = within(canvasElement).getByText('Sign in');
    const register = within(canvasElement).getByText('Register');
    expect(signIn).toBeInTheDocument();
    expect(register).toBeInTheDocument();
  },
};

export const Authenticated: Story = {
  args: {
    isSmallScreen: false,
  },
  decorators: [
    applicationConfig({
      providers: [{ provide: UserFacade, useValue: mockUserFacade(mockUser) }],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const signOut = within(canvasElement).getByText('Sign out');
    expect(signOut).not.toBeDisabled();
  },
};
