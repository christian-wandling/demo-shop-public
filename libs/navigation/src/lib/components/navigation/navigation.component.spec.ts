import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NavigationComponent } from './navigation.component';
import { NavigationService } from '../../services/navigation.service';
import { provideRouter, Router } from '@angular/router';
import { NavigationType } from '../../enums/navigation-type';
import { BrowserAnimationsModule, provideNoopAnimations } from '@angular/platform-browser/animations';
import { ProductFacade, ProductSearchComponent } from '@demo-shop/product';
import { Component } from '@angular/core';
import { CartIconComponent } from '@demo-shop/cart';
import { UserNavigationComponent } from './user-navigation/user-navigation.component';

describe('NavigationComponent', () => {
  let component: NavigationComponent;
  let fixture: ComponentFixture<NavigationComponent>;
  let navigationService: NavigationService;
  let router: Router;
  let productFacade: ProductFacade;

  @Component({ standalone: true, selector: 'lib-product-search', template: '' })
  class ProductSearchStubComponent {}

  @Component({ standalone: true, selector: 'lib-user-navigation', template: '' })
  class UserNavigationStubComponent {}

  @Component({ standalone: true, selector: 'lib-cart-icon', template: '' })
  class CartIconStubComponent {}

  const mockRouteItems = [
    {
      type: NavigationType.ROUTE,
      options: {
        route: 'route1',
        permissionStrategy: undefined,
        query: undefined,
      },
      label: 'Route1',
      order: 11,
    },
    {
      type: NavigationType.ROUTE,
      options: {
        route: 'route2',
        permissionStrategy: undefined,
        query: undefined,
      },
      label: 'Route2',
      order: 12,
    },
  ];

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [NavigationComponent, BrowserAnimationsModule],
      providers: [
        provideRouter([]),
        provideNoopAnimations(),
        {
          provide: NavigationService,
          useValue: {
            getFilteredItems: jest
              .fn()
              .mockImplementation(type => (type === NavigationType.ROUTE ? mockRouteItems : [])),
          },
        },
        { provide: ProductFacade, useValue: { setFilter: jest.fn() } },
        Router,
      ],
    })
      .overrideComponent(NavigationComponent, {
        remove: {
          imports: [ProductSearchComponent, UserNavigationComponent, CartIconComponent],
        },
        add: {
          imports: [ProductSearchStubComponent, UserNavigationStubComponent, CartIconStubComponent],
        },
      })
      .compileComponents();

    navigationService = TestBed.inject(NavigationService);
    router = TestBed.inject(Router);
    productFacade = TestBed.inject(ProductFacade);
    fixture = TestBed.createComponent(NavigationComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should initialize with default values', () => {
    expect(component.mobileMenuOpen()).toBeFalsy();
    expect(component.selectedMenuItem()).toBe('products');
  });

  it('should fetch route menu items on init', () => {
    expect(navigationService.getFilteredItems).toHaveBeenCalledWith(NavigationType.ROUTE);
    expect(component.menuItems).toEqual(mockRouteItems);
  });

  it('should navigate to correct category when setCategory is called', () => {
    jest.spyOn(router, 'navigateByUrl');
    const testCategory = 'test-category';

    component.setProductCategory(testCategory);

    expect(productFacade.setFilter).toHaveBeenCalledWith({ categories: testCategory });
    expect(router.navigateByUrl).toHaveBeenCalledWith('/products');
  });

  it('should close mobile menu on window resize', () => {
    component.mobileMenuOpen.set(true);

    window.dispatchEvent(new Event('resize'));
    fixture.detectChanges();

    expect(component.mobileMenuOpen()).toBeFalsy();
  });

  describe('snapshots', () => {
    it('should match snapshot with default state', () => {
      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should match snapshot with mobile menu open', () => {
      component.mobileMenuOpen.set(true);
      fixture.detectChanges();
      expect(fixture.nativeElement).toMatchSnapshot();
    });
  });
});
