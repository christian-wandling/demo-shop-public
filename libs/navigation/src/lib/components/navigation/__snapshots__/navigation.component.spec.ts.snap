// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationComponent snapshots should match snapshot with default state 1`] = `
<div
  class="bg-white"
  id="root4"
>
  
  <header
    class="relative bg-white"
  >
    <nav
      aria-label="Top"
      class="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8"
    >
      <div
        class="border-b border-gray-200"
      >
        <div
          class="flex h-16 items-center"
        >
          <button
            class="relative rounded-md bg-white p-2 text-gray-400 lg:hidden"
            type="button"
          >
            <span
              class="absolute -inset-0.5"
            />
            <span
              class="sr-only"
            >
              Open menu
            </span>
            <svg
              aria-hidden="true"
              class="h-6 w-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
            >
              <path
                d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <div
            class="ml-4 flex lg:ml-0 h-8 w-8 relative"
          >
            <a
              href="/"
              routerlink="/"
            >
              <span
                class="sr-only"
              >
                Demo Shop
              </span>
              <img
                alt="Demo shop logo"
                class="object-contain"
                fetchpriority="auto"
                fill=""
                loading="lazy"
                ng-img="true"
                ngsrc="icons/demo-shop.png"
                sizes="100vw"
                src="icons/demo-shop.png"
                style="position: absolute; width: 100%; height: 100%; inset: 0;"
              />
            </a>
          </div>
          <div
            class="hidden lg:ml-8 lg:block lg:self-stretch"
          >
            <div
              class="flex h-full space-x-8"
            >
              <div
                class="menu-item flex h-full space-x-8"
              >
                <a
                  class="menu-item-button flex items-center text-sm font-medium text-gray-700 hover:text-gray-800"
                  href="/route1"
                >
                   Route1 
                </a>
              </div>
              <div
                class="menu-item flex h-full space-x-8"
              >
                <a
                  class="menu-item-button flex items-center text-sm font-medium text-gray-700 hover:text-gray-800"
                  href="/route2"
                >
                   Route2 
                </a>
              </div>
              
            </div>
          </div>
          <div
            class="ml-auto flex items-center"
          >
            <lib-user-navigation
              class="hidden lg:flex lg:flex-1 lg:items-center lg:justify-end lg:space-x-6"
            />
            <lib-product-search />
            <lib-cart-icon />
          </div>
        </div>
      </div>
    </nav>
  </header>
</div>
`;

exports[`NavigationComponent snapshots should match snapshot with mobile menu open 1`] = `
<div
  class="bg-white"
  id="root5"
>
  <dialog
    aria-modal="true"
    class="relative z-40 lg:hidden ng-trigger ng-trigger-animateBackdrop ng-animating"
  >
    <div
      aria-hidden="true"
      class="fixed inset-0 bg-black bg-opacity-25"
    />
  </dialog>
  <div
    class="absolute h-screen z-40 flex w-full max-w-xs flex-col overflow-y-auto bg-white px-6 p-2 shadow-xl ng-trigger ng-trigger-animateSlideOver ng-animating"
  >
    <div
      class="sticky z-20 top-0 flex px-4 pb-2 pt-5"
    >
      <button
        class="relative opacity-90 -m-2 inline-flex items-center justify-center rounded-md p-2 text-gray-400 bg-white"
        type="button"
      >
        <span
          class="absolute -inset-0.5"
        />
        <span
          class="sr-only"
        >
          Close menu
        </span>
        <svg
          aria-hidden="true"
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 18L18 6M6 6l12 12"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
    <div
      class="flex-1 space-y-6 border-t border-gray-200 px-4 py-6"
    >
      <div
        class="flow-root"
        style=""
      >
        <a
          class="-m-2 block p-2 font-medium text-gray-700 hover:text-gray-800 text-xl"
          href="/route1"
        >
           Route1 
        </a>
      </div>
      <div
        class="flow-root"
        style=""
      >
        <a
          class="-m-2 block p-2 font-medium text-gray-700 hover:text-gray-800 text-xl"
          href="/route2"
        >
           Route2 
        </a>
      </div>
      
    </div>
    <lib-user-navigation
      class="lg:hidden space-y-6 border-t border-gray-200 px-4 py-6"
      issmallscreen=""
    />
  </div>
  
  <header
    class="relative bg-white"
  >
    <nav
      aria-label="Top"
      class="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8"
    >
      <div
        class="border-b border-gray-200"
      >
        <div
          class="flex h-16 items-center"
        >
          <button
            class="relative rounded-md bg-white p-2 text-gray-400 lg:hidden"
            type="button"
          >
            <span
              class="absolute -inset-0.5"
            />
            <span
              class="sr-only"
            >
              Open menu
            </span>
            <svg
              aria-hidden="true"
              class="h-6 w-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
            >
              <path
                d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <div
            class="ml-4 flex lg:ml-0 h-8 w-8 relative"
          >
            <a
              href="/"
              routerlink="/"
            >
              <span
                class="sr-only"
              >
                Demo Shop
              </span>
              <img
                alt="Demo shop logo"
                class="object-contain"
                fetchpriority="auto"
                fill=""
                loading="lazy"
                ng-img="true"
                ngsrc="icons/demo-shop.png"
                sizes="100vw"
                src="icons/demo-shop.png"
                style="position: absolute; width: 100%; height: 100%; inset: 0;"
              />
            </a>
          </div>
          <div
            class="hidden lg:ml-8 lg:block lg:self-stretch"
          >
            <div
              class="flex h-full space-x-8"
            >
              <div
                class="menu-item flex h-full space-x-8"
              >
                <a
                  class="menu-item-button flex items-center text-sm font-medium text-gray-700 hover:text-gray-800"
                  href="/route1"
                >
                   Route1 
                </a>
              </div>
              <div
                class="menu-item flex h-full space-x-8"
              >
                <a
                  class="menu-item-button flex items-center text-sm font-medium text-gray-700 hover:text-gray-800"
                  href="/route2"
                >
                   Route2 
                </a>
              </div>
              
            </div>
          </div>
          <div
            class="ml-auto flex items-center"
          >
            <lib-user-navigation
              class="hidden lg:flex lg:flex-1 lg:items-center lg:justify-end lg:space-x-6"
            />
            <lib-product-search />
            <lib-cart-icon />
          </div>
        </div>
      </div>
    </nav>
  </header>
</div>
`;
